design-proposals/network-device-capacity.md

```diff
@@ -0,0 +1,147 @@
+by the new user *domain-resources-requests-interface* input.
+
+> **Note**: While this solution focuses on network interfaces, the approach can be
+> applied to any device type which has dependencies on other devices.
+
+### Definition of Personas
+- VM/VMI administrator: Any user who can create VM and VMI objects.
+
+### User Stories
+- TBD
+
+### API Extensions
+The VMI Spec will be extended to include a new domain resource name: `interface`.
+It will be available as a domain-resources-requests entry.
+
+#### Validation
+- Should be immutable.
+- Should accept an integer value between 0 and 32.
+- Should be accepted only as `requests` (not as `limits`).
+- In case the value is less than the amount of interfaces defined already,
+  it has no effect.
+
+> **Note**: These are API level validations. The backend (e.g. hypervisor)
+> may have other validations/limitation.
+
+#### VMI Spec example
+```yaml
+---
+apiVersion: kubevirt.io/v1
+kind: VirtualMachineInstance
+metadata:
+  name: vmi-iface-request-example
+spec:
+  domain:
+    devices:
+      disks:
+        - disk:
+            bus: virtio
+          name: containerdisk
+      interfaces:
+        - masquerade: {}
+          name: default
+        - bridge: {}
+          name: secondary
+    resources:
+      requests:
+        interface: 5
+    rng: {}
+  networks:
+    - name: default
+      pod: {}
+    - multus:
+        networkName: some-nad
+      name: secondary 
+  volumes:
+    - containerDisk:
+        image: registry:5000/kubevirt/fedora-with-test-tooling-container-disk:devel
+      name: containerdisk
+```
+
+## Design Details
+The solution is taking advantage of the existing libvirt capability to auto-generate
+the needed configuration (e.g. controllers) when specifying network interfaces.
+
+The main change is in the logic of creating a domain in the virt-launcher.
+
+This is the main flow suggested:
+- Base on the resources-requests-interface and the current defined interfaces, generate
+  [interface-placeholders](#network-interface-placeholder) and add them to the domain spec.
+- Following the creation of the domain configuration, remove these placeholders
+  from he configuration and start the domain without them.
+- At this stage, the domain has been started without the additional interface devices
+  but with the additional configuration (e.g. controllers) they originally generated.
+
+#### Network Interface Placeholder
+The interface placeholder is a temporary dummy definition of an interface in the domain spec
+of libvirt. It is used to trigger libvirt to auto-generate dependent devices (e.g. PCIe controllers).
+
+After the domain configuration is constructed, these placeholders are removed.
+
+This is an example of such a placeholder:
+```xml
+<interface type='ethernet'>
+  <target dev='placeholder-1' managed='no'/>
+  <model type='virtio-non-transitional'/>
+  <mtu size='1500'/>
+  <rom enabled='no'/>
```

Comment:
why do you pass `mtu` and ` rom enabled='no'` om the placeholder? Can it be dropped?

Reply:
Done